File: frontend/src/types/prediction.ts
// ======================================================
// Bounding Box
// ======================================================

export interface BoundingBox {
  x1: number;
  y1: number;
  x2: number;
  y2: number;
}

// ======================================================
// Single Detection Result
// ======================================================

export interface DetectionResult {
  bbox: BoundingBox;

  detector_confidence: number;

  class_id: number;

  class_name: string;

  classification_confidence: number;

  description: string;

  recommended_action: string;
}

// ======================================================
// Live / Image Detection API Response
// ======================================================

export interface DetectionResponse {
  success?: boolean;

  prediction_type?: string;

  total_detections: number;

  detections: DetectionResult[];
}

// ======================================================
// Prediction History Item
// ======================================================

export interface PredictionResult {
  id?: string;

  user_id: string;

  prediction_type: "image" | "video" | "live";

  filename: string;

  total_detections: number;

  detections: DetectionResult[];

  created_at: string;
}

// ======================================================
// Prediction History API Response
// ======================================================

export interface PredictionHistoryResponse {
  success: boolean;

  total: number;

  predictions: PredictionResult[];
}

// ======================================================
// Delete Prediction Response
// ======================================================

export interface DeletePredictionResponse {
  success: boolean;

  message: string;
}

// ======================================================
// Video Prediction
// ======================================================

export interface VideoFrameResult {
  frame: number;

  detections: DetectionResult[];
}

export interface VideoPredictionResponse {
  success: boolean;

  prediction_type: "video";

  frames_processed: number;

  frames_with_detections: number;

  results: VideoFrameResult[];
}

File: backend/app/schemas/prediction.py
from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, Field


# ==========================================================
# BOUNDING BOX
# ==========================================================

class BoundingBox(BaseModel):
    x1: int
    y1: int
    x2: int
    y2: int


# ==========================================================
# SINGLE DETECTION
# ==========================================================

class Detection(BaseModel):
    bbox: Optional[BoundingBox] = None

    detector_confidence: float = Field(
        ...,
        ge=0.0,
        le=1.0
    )

    class_id: int

    class_name: str

    classification_confidence: float = Field(
        ...,
        ge=0.0,
        le=1.0
    )

    description: str

    recommended_action: str


# ==========================================================
# IMAGE / VIDEO / LIVE PREDICTION RESPONSE
# ==========================================================

class PredictionResponse(BaseModel):

    success: bool = True

    prediction_type: str

    total_detections: int

    detections: List[Detection]


# ==========================================================
# SINGLE PREDICTION HISTORY
# ==========================================================

class PredictionHistory(BaseModel):

    id: Optional[str] = None

    user_id: str

    prediction_type: str

    filename: str

    total_detections: int

    detections: List[Detection]

    created_at: datetime


# ==========================================================
# HISTORY LIST RESPONSE
# ==========================================================

class PredictionListResponse(BaseModel):

    total: int

    predictions: List[PredictionHistory]


# ==========================================================
# DELETE RESPONSE
# ==========================================================

class DeletePredictionResponse(BaseModel):

    success: bool

    message: str


# ==========================================================
# GENERIC RESPONSE
# ==========================================================

class ApiResponse(BaseModel):

    success: bool

    message: str